File: modules/napplet-host/ios/PublicAddressPolicy.swift
import Foundation

/// Conservative IP-literal policy for a future pinned native transport. This does not pin DNS.
public enum PublicAddressPolicy {
  /// Classifies an address already resolved by a transport, without any name lookup.
  public static func accepts(_ address: Data) -> Bool {
    let bytes = [UInt8](address)
    if bytes.count == 4 { return publicV4(bytes) }
    if bytes.count == 16 { return publicV6(bytes) }
    return false
  }

  /// True only for syntactically valid, globally routable unicast address literals.
  public static func isPublicLiteral(_ value: String) -> Bool {
    guard !value.isEmpty, !value.contains("%"), !value.contains("["), !value.contains("]") else { return false }
    if value.contains(":") {
      guard let bytes = parseV6(value) else { return false }
      return publicV6(bytes)
    }
    guard let bytes = parseV4(value) else { return false }
    return publicV4(bytes)
  }

  /// Reject the whole DNS result set if it is empty, malformed, or contains any non-public answer.
  public static func allPublicAnswers(_ answers: [String]) -> Bool {
    !answers.isEmpty && answers.allSatisfy(isPublicLiteral)
  }

  private static func parseV4(_ input: String) -> [UInt8]? {
    let parts = input.split(separator: ".", omittingEmptySubsequences: false)
    guard parts.count == 4 else { return nil }
    var result = [UInt8]()
    for part in parts {
      guard !part.isEmpty, part.count <= 3, !(part.count > 1 && part.first == "0"),
            part.allSatisfy({ $0 >= "0" && $0 <= "9" }), let number = UInt16(part), number <= 255 else { return nil }
      result.append(UInt8(number))
    }
    return result
  }

  private static func parseV6(_ input: String) -> [UInt8]? {
    if input.components(separatedBy: "::").count > 2 { return nil }
    var text = input
    if text.contains(".") {
      guard let colon = text.lastIndex(of: ":"), let v4 = parseV4(String(text[text.index(after: colon)...])) else { return nil }
      let tail = String(format: "%x:%x", (UInt16(v4[0]) << 8) | UInt16(v4[1]), (UInt16(v4[2]) << 8) | UInt16(v4[3]))
      text = String(text[...colon]) + tail
    }
    let compressed = text.contains("::")
    let halves = text.components(separatedBy: "::")
    let left = halves[0].isEmpty ? [] : halves[0].split(separator: ":", omittingEmptySubsequences: false).map(String.init)
    let right = halves.count == 1 || halves[1].isEmpty ? [] : halves[1].split(separator: ":", omittingEmptySubsequences: false).map(String.init)
    let count = left.count + right.count
    guard (compressed ? count < 8 : count == 8) else { return nil }
    let groups = left + Array(repeating: "0", count: compressed ? 8 - count : 0) + right
    guard groups.count == 8 else { return nil }
    var bytes = [UInt8]()
    for group in groups {
      guard !group.isEmpty, group.count <= 4, group.utf8.allSatisfy({ (48...57).contains($0) || (65...70).contains($0) || (97...102).contains($0) }),
            let number = UInt16(group, radix: 16) else { return nil }
      bytes.append(UInt8(number >> 8))
      bytes.append(UInt8(number & 0xff))
    }
    return bytes
  }

  private static func publicV4(_ a: [UInt8]) -> Bool {
    !(hasPrefix(a, [0], 8) || hasPrefix(a, [10], 8) || hasPrefix(a, [100, 64], 10) || hasPrefix(a, [127], 8) ||
      hasPrefix(a, [169, 254], 16) || hasPrefix(a, [172, 16], 12) || hasPrefix(a, [192, 0, 0], 24) ||
      hasPrefix(a, [192, 0, 2], 24) || hasPrefix(a, [192, 88, 99], 24) || hasPrefix(a, [192, 168], 16) ||
      hasPrefix(a, [198, 18], 15) || hasPrefix(a, [198, 51, 100], 24) || hasPrefix(a, [203, 0, 113], 24) ||
      hasPrefix(a, [224], 4) || hasPrefix(a, [240], 4))
  }

  private static func publicV6(_ a: [UInt8]) -> Bool {
    guard a.count == 16, (a[0] & 0xe0) == 0x20 else { return false } // Only 2000::/3.
    if a[0] == 0x20 && a[1] == 0x01 && a[2] <= 0x01 { return false } // 2001::/23 special use.
    if a[0] == 0x20 && a[1] == 0x01 && a[2] == 0x0d && a[3] == 0xb8 { return false } // Documentation.
    if a[0] == 0x20 && a[1] == 0x02 { return false } // 6to4.
    if a[0] == 0x3f && a[1] == 0xff && (a[2] & 0xf0) == 0 { return false } // 3fff::/20 documentation.
    // IPv4-mapped forms are rejected even when the embedded IPv4 address is public.
    return !a.prefix(10).allSatisfy({ $0 == 0 }) || a[10] != 0xff || a[11] != 0xff
  }

  private static func hasPrefix(_ address: [UInt8], _ prefix: [UInt8], _ bits: Int) -> Bool {
    var remaining = bits
    for index in prefix.indices {
      let count = min(8, remaining)
      if count == 0 { break }
      let mask = UInt8(0xff) << (8 - count)
      if (address[index] & mask) != (prefix[index] & mask) { return false }
      remaining -= count
    }
    return true
  }
}

/// Descriptive name for native published-artifact transports.
public typealias PublishedPublicAddressPolicy = PublicAddressPolicy

File: modules/napplet-host/android/src/main/java/org/nostrocket/hypergolic/host/PublicAddressPolicy.java
package org.nostrocket.hypergolic.host;

import java.util.List;

/** Conservative IP-literal policy for a future pinned native transport. This does not pin DNS. */
public final class PublicAddressPolicy {
  private PublicAddressPolicy() {}

  /** Classifies an address already resolved by a transport, without any name lookup. */
  public static boolean accepts(byte[] address) {
    if (address == null) return false;
    if (address.length == 4) return publicV4(address);
    if (address.length == 16) return publicV6(address);
    return false;
  }

  /** True only for syntactically valid, globally routable unicast address literals. */
  public static boolean isPublicLiteral(String value) {
    if (value == null || value.isEmpty() || value.indexOf('%') >= 0 || value.indexOf('[') >= 0 || value.indexOf(']') >= 0) return false;
    if (value.indexOf(':') >= 0) {
      byte[] address = parseV6(value);
      return address != null && publicV6(address);
    }
    byte[] address = parseV4(value);
    return address != null && publicV4(address);
  }

  /** Reject the whole DNS result set if it is empty, malformed, or contains any non-public answer. */
  public static boolean allPublicAnswers(List<String> answers) {
    if (answers == null || answers.isEmpty()) return false;
    for (String answer : answers) if (!isPublicLiteral(answer)) return false;
    return true;
  }

  private static byte[] parseV4(String input) {
    String[] parts = input.split("\\.", -1);
    if (parts.length != 4) return null;
    byte[] out = new byte[4];
    for (int i = 0; i < 4; i++) {
      String part = parts[i];
      if (part.isEmpty() || part.length() > 3 || (part.length() > 1 && part.charAt(0) == '0')) return null;
      int n = 0;
      for (int j = 0; j < part.length(); j++) {
        char c = part.charAt(j);
        if (c < '0' || c > '9') return null;
        n = n * 10 + c - '0';
      }
      if (n > 255) return null;
      out[i] = (byte) n;
    }
    return out;
  }

  private static byte[] parseV6(String input) {
    if (input.indexOf("::") != input.lastIndexOf("::")) return null;
    String text = input;
    int dot = text.indexOf('.');
    if (dot >= 0) {
      int colon = text.lastIndexOf(':');
      if (colon < 0) return null;
      byte[] v4 = parseV4(text.substring(colon + 1));
      if (v4 == null) return null;
      int first = ((v4[0] & 0xff) << 8) | (v4[1] & 0xff);
      int second = ((v4[2] & 0xff) << 8) | (v4[3] & 0xff);
      text = text.substring(0, colon + 1) + Integer.toHexString(first) + ":" + Integer.toHexString(second);
    }
    boolean compressed = text.contains("::");
    String[] halves = text.split("::", -1);
    String[] left = halves[0].isEmpty() ? new String[0] : halves[0].split(":", -1);
    String[] right = halves.length == 1 || halves[1].isEmpty() ? new String[0] : halves[1].split(":", -1);
    int count = left.length + right.length;
    if ((compressed && count >= 8) || (!compressed && count != 8)) return null;
    int zeros = compressed ? 8 - count : 0;
    byte[] out = new byte[16];
    int index = 0;
    for (String group : left) { if (!putGroup(group, out, index++)) return null; }
    index += zeros;
    for (String group : right) { if (!putGroup(group, out, index++)) return null; }
    return index == 8 ? out : null;
  }

  private static boolean putGroup(String group, byte[] out, int index) {
    if (group.isEmpty() || group.length() > 4 || index < 0 || index > 7) return false;
    int n = 0;
    for (int i = 0; i < group.length(); i++) {
      int d = Character.digit(group.charAt(i), 16);
      if (d < 0 || group.charAt(i) > 127) return false;
      n = (n << 4) | d;
    }
    out[index * 2] = (byte) (n >>> 8);
    out[index * 2 + 1] = (byte) n;
    return true;
  }

  private static boolean publicV4(byte[] a) {
    return !(prefix(a, 0, 8) || prefix(a, 10, 8) || prefix2(a, 100, 64, 10) || prefix(a, 127, 8) ||
        prefix2(a, 169, 254, 16) || prefix2(a, 172, 16, 12) || prefix3(a, 192, 0, 0) ||
        prefix3(a, 192, 0, 2) || prefix3(a, 192, 88, 99) || prefix2(a, 192, 168, 16) ||
        prefix2(a, 198, 18, 15) || prefix3(a, 198, 51, 100) || prefix3(a, 203, 0, 113) ||
        prefix(a, 224, 4) || prefix(a, 240, 4));
  }

  private static boolean publicV6(byte[] a) {
    // Only 2000::/3 global-unicast space is considered. This excludes unspecified,
    // loopback, ULA, link-local, multicast, NAT64, IPv4-compatible and most special use.
    if ((a[0] & 0xe0) != 0x20) return false;
    if ((a[0] & 0xff) == 0x20 && (a[1] & 0xff) == 0x01 && (a[2] & 0xff) <= 0x01) return false; // 2001::/23
    if ((a[0] & 0xff) == 0x20 && (a[1] & 0xff) == 0x01 && (a[2] & 0xff) == 0x0d && (a[3] & 0xff) == 0xb8) return false; // documentation
    if ((a[0] & 0xff) == 0x20 && (a[1] & 0xff) == 0x02) return false; // 6to4
    if ((a[0] & 0xff) == 0x3f && (a[1] & 0xff) == 0xff && (a[2] & 0xf0) == 0) return false; // 3fff::/20 documentation
    // IPv4-mapped forms are rejected even when the embedded IPv4 address is public.
    for (int i = 0; i < 10; i++) if (a[i] != 0) return true;
    return a[10] != (byte) 0xff || a[11] != (byte) 0xff;
  }

  private static boolean prefix(byte[] a, int first, int bits) {
    int mask = (0xff << (8 - bits)) & 0xff;
    return (a[0] & mask) == (first & mask);
  }
  private static boolean prefix2(byte[] a, int x, int y, int bits) {
    int value = ((a[0] & 0xff) << 8) | (a[1] & 0xff);
    return value >>> (16 - bits) == ((x << 8) | y) >>> (16 - bits);
  }
  private static boolean prefix3(byte[] a, int x, int y, int z) {
    return (a[0] & 0xff) == x && (a[1] & 0xff) == y && (a[2] & 0xff) == z;
  }
}
